README
NAME
     Video::Delay - Perl classes for delays between frames generation.

SEE ALSO
    Video::Delay::Array
        Video::Delay class for predefined list of delays.

    Video::Delay::Const
        Video::Delay class for constant delay.

    Video::Delay::Func
        Video::Delay class for delays defined by math function.

REPOSITORY
    <https://github.com/michal-josef-spacek/Video-Delay>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
     © 2012-2020 Michal Josef Špaček
     BSD 2-Clause License

VERSION
    0.08



Changes
0.08
 [Common]
 - Update .travis.yml to reality.

0.07 2020-01-19T14:56:23+01:00
 [Common]
 - Fix Makefile.PL fails with no '.' in @INC.
 - Update Module::Install to 1.19 version.
 - Update author github username.
 - Update author name.
 - Update copyright years.

0.06 2015-09-16T20:02:25+02:00
 [Common]
 - Move author tests to xt/ directory.

0.05 2015-09-15T21:30:37+02:00
 [Common]
 - Improve LICENSE AND COPYRIGHT sections in doc.
 - Improve SEE ALSO sections in doc.
 - Update Module::Install to 1.16 version.
 - Update copyright years.

0.04 2014-11-03T12:55:54+01:00
 [Video::Delay::Func]
 - Fix test (precision).

0.03 2014-11-03T07:29:41+01:00
 [Video::Delay::Array]
 - Add tests.

 [Video::Delay::Const]
 - Add tests.

 [Video::Delay::Func]
 - Add callback for 'func' parameter.
 - Add EXAMPLE section to doc.
 - Add error check for 'func' parameter.
 - Add tests.
 - Improve error message.
 - Rewrite default value of 'func' to subroutine reference.

0.02 2014-10-31T08:04:17+01:00
 - Add README file.
 - Add Video::Delay module.

0.01 2014-10-31T07:56:31+01:00
 - First version.


Delay/Func.pm
package Video::Delay::Func;

use strict;
use warnings;

use Class::Utils qw(set_params);
use English qw(-no_match_vars);
use Error::Pure qw(err);

our $VERSION = 0.08;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Math function.
	$self->{'func'} = sub {
		my $t = shift;
		return 1000 * sin($t);
	};

	# Counter increment.
	$self->{'incr'} = 0.1;

	# Process params.
	set_params($self, @params);

	# Counter.
	$self->{'counter'} = 0;

	# Check 'func' parameters.
	if (ref $self->{'func'} ne '' && ref $self->{'func'} ne 'CODE') {
		err "Parameter 'func' must be scalar or code.";
	}

	# Object.
	return $self;
}

# Get delay.
sub delay {
	my $self = shift;

	# Counter.
	$self->{'counter'} += $self->{'incr'};

	# Function string.
	my $ret;
	if (ref $self->{'func'} eq '') {
		my $input = $self->{'func'};
		my $c = $self->{'counter'};
		$input =~ s/t/$c/g;

		# Eval.
		$ret = eval $input;
		if ($EVAL_ERROR) {
			err 'Error in function.',
				'Error', $EVAL_ERROR;
		}

	# Callback.
	} else {
		$ret = $self->{'func'}->($self->{'counter'});
	}

	return $ret;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Video::Delay::Func - Video::Delay class for delays defined by math function.

=head1 SYNOPSIS

 use Video::Delay::Func;

 my $obj = Video::Delay::Func->new(%parameters);
 my $delay = $obj->delay;

=head1 METHODS

=over 8

=item C<new(%parameters)>

 Constructor

=over 8

=item * C<func>

 Math function in string with 't' time variable.
 Or callback with one input argument as time variable.
 Default value is reference to subroutine:
   sub {
     my $t = shift;
     return 1000 * sin($t);
   };

=item * C<incr>

 Counter increment.
 Default value is 0.1.

=back

=item C<delay()>

 Returns delay defined by function in miliseconds.
 Time si affected by 'incr' increment parameter.

=back

=head1 ERRORS

 new():
         Parameter 'func' must be scalar or code.
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

 delay():
         Error in function.
                 Error, %s

=head1 EXAMPLE1

 use strict;
 use warnings;

 use Video::Delay::Func;

 # Object.
 my $obj = Video::Delay::Func->new(
         'func' => sub {
                 my $t = shift;
                 return 1000 * sin($t);
         },
         'incr' => 0.1,
 );

 # Print delay.
 print $obj->delay."\n";
 print $obj->delay."\n";
 print $obj->delay."\n";

 # Output:
 # 99.8334166468282
 # 198.669330795061
 # 295.52020666134

=head1 EXAMPLE2

 use strict;
 use warnings;

 use Video::Delay::Func;

 # Object.
 my $obj = Video::Delay::Func->new(
         'func' => '1000 * sin(t)',
         'incr' => 0.1,
 );

 # Print delay.
 print $obj->delay."\n";
 print $obj->delay."\n";
 print $obj->delay."\n";

 # Output:
 # 99.8334166468282
 # 198.669330795061
 # 295.52020666134

=head1 DEPENDENCIES

L<Class::Utils>,
L<English>,
L<Error::Pure>.

=head1 SEE ALSO

=over

=item L<Video::Delay>

Perl classes for delays between frames generation.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Video-Delay>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

 © 2012-2020 Michal Josef Špaček
 BSD 2-Clause License

=head1 VERSION

0.08

=cut


Delay/Array.pm
package Video::Delay::Array;

use strict;
use warnings;

use Class::Utils qw(set_params);

our $VERSION = 0.08;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Array.
	$self->{'array'} = [1000, 2000, 3000];

	# Loop.
	$self->{'loop'} = 1;

	# Process params.
	set_params($self, @params);

	# Index.
	$self->{'index'} = 0;

	# Object.
	return $self;
}

# Get delay.
sub delay {
	my $self = shift;

	# Return value.
	my $ret;
	if ($self->{'index'} <= @{$self->{'array'}} - 1) {
		$ret = $self->{'array'}->[$self->{'index'}];
	}

	# Increment.
	$self->{'index'}++;
	if ($self->{'loop'} && $self->{'index'} > @{$self->{'array'}} - 1) {
		$self->{'index'} = 0;
	}

	# Value.
	return $ret;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Video::Delay::Array - Video::Delay class for predefined list of delays.

=head1 SYNOPSIS

 use Video::Delay::Array;

 my $obj = Video::Delay::Array->new(%parameters);
 my $delay = $obj->delay;

=head1 METHODS

=over 8

=item C<new(%parameters)>

 Constructor

=over 8

=item * C<array>

 Array of predefined delays in miliseconds.
 Default value is [1000, 2000, 3000].

=item * C<loop>

 Loop flag.
 Default value is 1.

=back

=item C<delay()>

 Returns delay from array defined by 'array' parameter in miliseconds.
 Affected by 'loop' parameter.

=back

=head1 ERRORS

 new():
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

=head1 EXAMPLE

 use strict;
 use warnings;

 use Video::Delay::Array;

 # Object.
 my $obj = Video::Delay::Array->new(
         'array' => [1000, 2000],
         'loop' => 1,
 );

 # Print delay.
 print $obj->delay."\n";
 print $obj->delay."\n";
 print $obj->delay."\n";

 # Output:
 # 1000
 # 2000
 # 1000

=head1 DEPENDENCIES

L<Class::Utils>.

=head1 SEE ALSO

=over

=item L<Video::Delay>

Perl classes for delays between frames generation.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Video-Delay>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

 © 2012-2020 Michal Josef Špaček
 BSD 2-Clause License

=head1 VERSION

0.08

=cut


Delay/Const.pm
package Video::Delay::Const;

use strict;
use warnings;

use Class::Utils qw(set_params);

our $VERSION = 0.08;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Constant.
	$self->{'const'} = 1000;

	# Process params.
	set_params($self, @params);

	# Object.
	return $self;
}

# Get delay.
sub delay {
	my $self = shift;
	return $self->{'const'};
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Video::Delay::Const - Video::Delay class for constant delay.

=head1 SYNOPSIS

 use Video::Delay::Const;

 my $obj = Video::Delay::Const->new(%parameters);
 my $delay = $obj->delay;

=head1 METHODS

=over 8

=item C<new(%parameters)>

 Constructor

=over 8

=item * C<const>

 Constant delay in miliseconds.
 Default value is 1000.

=back

=item C<delay()>

 Returns constant delay defined by 'const' parameter in miliseconds.

=back

=head1 ERRORS

 new():
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

=head1 EXAMPLE

 use strict;
 use warnings;

 use Video::Delay::Const;

 # Object.
 my $obj = Video::Delay::Const->new(
         'const' => 1000,
 );

 # Print delay.
 print $obj->delay."\n";
 print $obj->delay."\n";
 print $obj->delay."\n";

 # Output:
 # 1000
 # 1000
 # 1000

=head1 DEPENDENCIES

L<Class::Utils>.

=head1 SEE ALSO

=over

=item L<Video::Delay>

Perl classes for delays between frames generation.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Video-Delay>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

 © 2012-2020 Michal Josef Špaček
 BSD 2-Clause License

=head1 VERSION

0.08

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Perl class for delays between frames generation.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
author_requires 'English' => 0;
author_requires 'File::Object' => 0.05;
license 'bsd';
name 'Video-Delay';
readme_from 'Delay.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0.03;
requires 'English' => 0;
requires 'Error::Pure' => 0.11;
requires 'perl' => '5.6.2';
resources 'bugtracker' => 'https://rt.cpan.org/Public/Dist/Display.html'.
	'?Name=Video-Delay';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Video-Delay';
resources 'repository' => 'git://github.com/michal-josef-spacek/Video-Delay';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0.17;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.08';

# Run.
WriteAll();
